client: more suggestions

azat opened the pull request:
Added the following to suggestion list:
- error code names
- events
- metrics
- macros
- storage policies

Changelog category (leave one):
- Not for changelog (changelog entry is not required)

azat committed:
client/suggest: add error code names

alexey-milovidov reviewed (commented):
Ok, but:

- we don't have suggest for three other tables with metrics;
- it will be affected by "zero on empty" setting.

azat committed:
client/suggest: set system_events_show_zero_values for query

azat committed:
client/suggest: add other metrics

azat committed:
client/suggest: add macros

azat committed:
client/suggest: add policy

azat commented:
>it will be affected by "zero on empty" setting.

Indeed, nice catch!

azat commented:
>AST fuzzer (ASan) — Received signal -3

https://clickhouse-test-reports.s3.yandex.net/19584/5f494c898af82cd46d68411696fa2d2ca0ef6184/fuzzer_asan/report.html#fail1

```
==86==WARNING: failed to fork (errno 12)
==86==WARNING: failed to fork (errno 12)
...
SUMMARY: AddressSanitizer: out-of-memory (/workspace/clickhouse+0xa908e79) 
```

alexey-milovidov commented:
Could you please add a test (with `expect`) that suggestions work?

azat committed:
tests: add a test for clickhouse-client autocompletion

v2: Increase timeout for 01676_clickhouse_client_autocomplete
    https://github.com/ClickHouse/ClickHouse/pull/19584#discussion_r565727175

v3: Disable 01676_clickhouse_client_autocomplete in unbundled build (arcadia)

    autocomplete does not have to work fully unbundled build (since it lack
    of replxx).

    Similar to bd523a0aff1e1355300febcddadced05f393a15f

v4: set expect timeout back to 1 and increase total timeout to 20 sec
v4: set expect timeout back to 3 and increase total timeout to 22 (3*X+1) sec

azat commented:
>Could you please add a test (with expect) that suggestions work?

OK, done (Hope that it will pass in arcadia, I tried not to use some exotic things)

alexey-milovidov commented on tests/queries/0_stateless/01676_clickhouse_client_autocomplete.sh:
```diff
@@ -0,0 +1,114 @@
+while {\$is_done == 0} {
+    send -- "\\t"
+    expect {
+        "$compword_begin$compword_end" {
```
Maybe just `$compword_end`

It completes instantly in non-ambiguous case, and displays colored list of variants in other cases.
Your example looks like to fail in both cases (maybe you are using readline that does not have colored suggestions).

azat replied:
>Maybe just $compword_end

Initially I wrote like this, but this not work correctly for cases like `foobar_aaaaaaa`, since last 3 bytes is the same.

> It completes instantly in non-ambiguous case, and displays colored list of variants in other cases.

Yes, but firstly it completes the base part, i.e. `forc` -> `force_` w/o coloring and display list of suggestions

> Your example looks like to fail in both cases

That's how it's meant to be (stricter is better), and the variants for completion was selected that way.
I think that the failures was due to too small timeout, increased, let's see.

> (maybe you are using readline that does not have colored suggestions).

Too big difference, I stopped doing this.

azat commented:
>Fast test — fail: 4, passed: 569, skipped: 61. Other tests won't be run before fast test will be fixed.

https://clickhouse-test-reports.s3.yandex.net/19584/7fb618e00f133e8d51b0d371e0781820520f7595/fast_test.html#fail1

Some CI issues?

```
2021.01.28 07:34:41.906245 [ 22407 ] {} <Trace> BaseDaemon: Received signal 15
```

Rebased against upstream.

alexey-milovidov commented:
It was an issue that fast test stopped to restart parallel tests, but I don't know the current status, need to ask @alesapin

alexey-milovidov commented:
It is Ok.

azat commented:
>Yandex synchronization check (only for Yandex employees)

Lack of autocompletion, will disable the test

alexey-milovidov commented:
Ok, there was also one failure of the test for roaring bitmaps that I also disabled for Arcadia in master.

azat commented:
>there was also one failure of the test for roaring bitmaps that I also disabled for Arcadia in master.

I've rebased against upstream too, so the check should pass

alexey-milovidov commented:
Ok but the test does not look good enough... If you cannot fix it let's proceed without the test (give up).

azat commented:
>but the test does not look good enough

Let's try one more time...

azat committed:
Make system.asynchronous_metrics available just after start

azat commented:
>Functional stateless tests flaky check (address) — fail: 0, passed: 72

Finally, passed.

The problem was in `system.asynchronous_metrics`, and that fact that they was populated only after `asynchronous_metrics_update_period_s` seconds from start of the server

alexey-milovidov committed:
Update Suggest.cpp

alexey-milovidov committed:
Merge branch 'master' into azat-client-complete-error-codes

alexey-milovidov commented:
Please look at the GRPC tests. They are green in master.

azat committed:
Fix SIGSEGV due to accessing GRPCServer::currentConnections() before initialization

This PR makes asynchronous metrics available just after start, and this
breaks GRPCServer, since it requires the server be started before
accessing currentConnections().

However it is possible to trigger the same SIGSEGV even without that
patch, with some timeouts during previous server initilizations and
small asynchronous_metrics_update_period_s.

Fix this by creating GRPCServer::Runner in the ctor.

Stacktrace:
    26.842505 [ 7 ] {} <Information> Application: Listening for MySQL compatibility protocol: 0.0.0.0:9004
    26.842562 [ 7 ] {} <Information> Application: Listening for gRPC protocol: 0.0.0.0:9100
    26.842600 [ 7 ] {} <Debug> AsynchronousMetrics: MemoryTracking: was 350.72 KiB, peak 5.25 MiB, will set to 232.63 MiB (RSS), difference: 232.29 MiB
    26.842834 [ 8 ] {} <Trace> BaseDaemon: Received signal 11
    26.843014 [ 39 ] {} <Fatal> BaseDaemon: ########################################
    26.843055 [ 39 ] {} <Fatal> BaseDaemon: (version 21.2.1.5858, build id: B27D5550AC34F9091BC4437D8021B752EDB34FBB) (from thread 7) (no query) Received signal Segmentation fault (11)
    26.843085 [ 39 ] {} <Fatal> BaseDaemon: Address: 0x78 Access: read. Address not mapped to object.
    26.843104 [ 39 ] {} <Fatal> BaseDaemon: Stack trace: 0x7fe8c4e6afc4 0x13683cf1 0xf52efe5 0xe7c9137 0x85596e1 0x8553635 0x11b7a313 0x8545ebc 0x8544b25 0x84e1cbe 0x7fe8c4c940b3 0x84ac22e
    26.843136 [ 39 ] {} <Fatal> BaseDaemon: 2. pthread_mutex_lock @ 0xbfc4 in /usr/lib/x86_64-linux-gnu/libpthread-2.31.so
    26.843162 [ 39 ] {} <Fatal> BaseDaemon: 3. std::__1::mutex::lock() @ 0x13683cf1 in ?
    26.843188 [ 39 ] {} <Fatal> BaseDaemon: 4. DB::GRPCServer::currentConnections() const @ 0xf52efe5 in /usr/bin/clickhouse
    26.843204 [ 39 ] {} <Fatal> BaseDaemon: 5. DB::AsynchronousMetrics::update() @ 0xe7c9137 in /usr/bin/clickhouse
    26.843228 [ 39 ] {} <Fatal> BaseDaemon: 6. DB::AsynchronousMetrics::start() @ 0x85596e1 in /usr/bin/clickhouse
    26.843248 [ 39 ] {} <Fatal> BaseDaemon: 7. DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) @ 0x8553635 in /usr/bin/clickhouse
    26.843270 [ 39 ] {} <Fatal> BaseDaemon: 8. Poco::Util::Application::run() @ 0x11b7a313 in /usr/bin/clickhouse
    26.843284 [ 39 ] {} <Fatal> BaseDaemon: 9. DB::Server::run() @ 0x8545ebc in /usr/bin/clickhouse
    26.843299 [ 39 ] {} <Fatal> BaseDaemon: 10. mainEntryClickHouseServer(int, char**) @ 0x8544b25 in /usr/bin/clickhouse
    26.843313 [ 39 ] {} <Fatal> BaseDaemon: 11. main @ 0x84e1cbe in /usr/bin/clickhouse
    26.843331 [ 39 ] {} <Fatal> BaseDaemon: 12. __libc_start_main @ 0x270b3 in /usr/lib/x86_64-linux-gnu/libc-2.31.so
    26.843346 [ 39 ] {} <Fatal> BaseDaemon: 13. _start @ 0x84ac22e in /usr/bin/clickhouse

azat commented:
>Please look at the GRPC tests. They are green in master.

978f23e3437d0d0eb64727e87ffe1b80289c9fc2

alexey-milovidov commented:
> AST fuzzer (UBSan) — ../src/Functions/DivisionUtils.h:90:40: runtime error: nan is outside the range of re

#19876

alexey-milovidov commented:
Could you please take a look at the `test_materialize_mysql_database`?

alexey-milovidov commented:
>  AST fuzzer (debug) — Logical error: 'Bad cast from type DB::ColumnVector<unsigned long> to DB::ColumnTuple'.

@kitaisreal This looks like a recent issue.

azat commented:
>@azat Could you please take a look at the test_materialize_mysql_database?

@alexey-milovidov The test is flaky, apparent part of data came after SELECT query:

```
2021.01.31 14:04:43.615618 [ 54 ] {} <Debug> executeQuery: (internal) /*Materialize MySQL step 1: execute dump data*/ INSERT INTO t(a, b, _sign, _version) VALUES
2021.01.31 14:04:43.616453 [ 54 ] {} <Trace> db.t (fc55a47e-1cae-4f41-9a13-91ccde5a73bb): Renaming temporary part tmp_insert_0_1_1_0 to 0_1_1_0.
...
2021.01.31 14:04:43.786823 [ 12 ] {4a0d1f80-4a96-4809-a84e-d55d7cd25cdd} <Debug> executeQuery: (from 172.18.0.1:45162, using production parser) SELECT count(_version) FROM db.t FORMAT TSV
...
2021.01.31 14:04:44.566037 [ 54 ] {} <Debug> executeQuery: (internal) /*Materialize MySQL step 1: execute dump data*/ INSERT INTO t(a, b, _sign, _version) VALUES
2021.01.31 14:04:44.566943 [ 54 ] {} <Trace> db.t (fc55a47e-1cae-4f41-9a13-91ccde5a73bb): Renaming temporary part tmp_insert_0_2_2_0 to 0_2_2_0.
...
2021.01.31 14:04:52.343911 [ 54 ] {} <Error> MaterializeMySQLSyncThread: Code: 1002, e.displayText() = DB::Exception: Master maybe lost, Stack trace (when copying this message, always include the lines below):
```

@zhang2014 what do you think about retrying those queries few times (or until they will match)?

azat committed:
Merge remote-tracking branch 'upstream/master' into client-complete-error-codes

alexey-milovidov commented:
> AST fuzzer (UBSan) — ../src/Functions/DivisionUtils.h:103:40: runtime error: 9.22337e+22 is outside the r

#19876

alexey-milovidov commented:
Need to debug integration test: `test_dictionaries_dependency_xml`.

azat commented:
>Need to debug integration test: test_dictionaries_dependency_xml.

Not related, but #19971 (that PR also includes https://github.com/ClickHouse/ClickHouse/pull/19584#issuecomment-771024679)
